In [12]:
#Importing libraries: pandas, NumPy, and os
import pandas as pd
import numpy as np
import os

In [14]:
#Creating path
path = r'C:\Users\isaac\Documents\CareerFoundry\2. Data Inmersion\4. Python Fundamentals for Data Analysts\07-2024 Instacart Basket Analysis\02 Data\Prepared Data\4.3_orders_products'


In [16]:
#Importing data 'orders_wrangled.csv' using path_join function
df_ords = pd.read_csv(os.path.join(path, 'orders_wrangled.csv'), index_col = False)

#Importing data products.csv using path_join function
df_prods = pd.read_csv(os.path.join(path, 'products.csv'), index_col = False)       

------------------------------------------------------------------------------------------------------------------------------------------

• Data Consistency Checks:

In [12]:
#Looking for descriptive statistics in 'df_prods' data frame:

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [19]:
#Checking for mixed data types on 'df_prods'

print(df_prods.dtypes)


product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object


In [23]:
#Checking for missing values

print(df_prods.isnull().sum())


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64


In [36]:
#Filtering out rows with missing in 'df_prods'

df_prods_clean = df_prods.dropna(subset=['product_name'])

#Checking rows number on new data frame 'df_prods_clean':
df_prods_clean.shape


(49677, 5)

In [41]:
#Removing any duplicate if any from 'df_prods_clean':

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [45]:
#Exporting 'df_prods_clean_no_dups' as 'products_checked.csv'

file_path = r'C:\Users\isaac\Documents\CareerFoundry\2. Data Inmersion\4. Python Fundamentals for Data Analysts\07-2024 Instacart Basket Analysis\02 Data\Prepared Data\4.3_orders_products\products_checked.csv'

products_checked = df_prods_clean_no_dups
products_checked.to_csv(file_path, index=False)



-----------------------------------------------------------------------------------------------------------------------------------------------------

• Data consistency checks on 'df_ords' data frame

In [20]:
#Looking for descriptive statistics in 'orders_wrangled.csv' data frame:

df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

• The count for days_since_prior_order is approximately 3.21 million, while the other columns have a count of approximately 3.42 million. This indicates that there are missing values in the days_since_prior_order column

• Column 'order_number' has a large range with a maximum value of 100, which could be an outlier or represent a small group of very frequent shoppers.







In [26]:
#Checking for mixed-type data in 'df_ords' dataframe

print(df_ords.dtypes)

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object


No mixed-type variables found

• Approximately 6.14% of the data in the days_since_prior_order column is missing. Therefore, I will proceed to replace the missing values with the mean

In [47]:
# Calculating the mean of the 'days_since_prior_order' column, excluding missing values
mean_value = df_ords['days_since_prior_order'].mean()

# Imputing missing values in 'days_since_prior_order' by creating a new dataframe
df_ords_imputed = df_ords.copy()
df_ords_imputed['days_since_prior_order'] = df_ords['days_since_prior_order'].fillna(mean_value)

# Checking the the new dataframe shape and missing values
print(df_ords_imputed.shape)
print(df_ords_imputed.isnull().sum())

(3421083, 7)
order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64


In [49]:
# Checking for duplicates
duplicate_records = df_ords_imputed.duplicated()
print(f"Number of duplicate records: {duplicate_records.sum()}")

Number of duplicate records: 0


• Final DataFrame Check:

In [52]:
# Check the dataframe information
print(df_ords_imputed.info())

# Display the first few rows of the cleaned dataframe
print(df_ords_imputed.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB
None
   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_ord

• Changing 'df_ords_imputed' name to then extract it:

In [57]:
#Renaming data frame 'df_ords_imputed'
df_ords_imputed = 

path_ords = r'C:\Users\isaac\Documents\CareerFoundry\2. Data Inmersion\4. Python Fundamentals for Data Analysts\07-2024 Instacart Basket Analysis\02 Data\Prepared Data\4.3_orders_products\orders_checked.csv'

#Exporting it:
orders_checked.to_csv(path_ords, index=False)

NameError: name 'orders_checked' is not defined